### ISSUE TITLE
Can't specify snr to balance multiple background tracks

### ISSUE THREAD

**Issue by bensteers** [justinsalamon/scaper#47]
I'm not sure why snr, pitch_shift, and time_stretch shouldn't work for bg events as well.

**Comment by justinsalamon** [comment]
SNR is a relative measure. All foreground event SNR is set relative to the background track, it doesn't really make sense for the background track to have an SNR in a vaccum.

Pitch shifting and time stretching is another story, and indeed we could support these values for the background track.

**Comment by bensteers** [comment]
I realize that SNR for background events is redundant when using a single background event (as you can control the level with ref_db), but if you want to add multiple background tracks, like for example if you have park noise and you want to mix in some birds, you can't control their relative levels. I suppose you could just set it as a foreground event with event_time=0 and event_duration=duration, but that ignores the semantics of background/foreground.

If you are adamant about enforcing only the use of a single background, you could raise an exception if multiple backgrounds are added, but I think multiple backgrounds can be convenient, but it's an issue if you can't balance them.

I agree that you shouldn't be required to specify an SNR for a background event and if nothing is specified it defaults to 0.

**Comment by justinsalamon** [comment]
I think there are two separate issues here that are being confounded:
1. Adding support for multiple backgrounds
2. How to add this support

Regarding (1), I am actually open to this idea - I can imagine people wanting to create a complex background scene out of multiple input files.

Regarding (2), the problem is there is no such thing as SNR for a single recording - SNR is only defined as a relative thing. So if you add multiple background files, you can't give each one it's own SNR because what would that SNR be relative to? It doesn't make sense. As I see it there are 2 alternatives:

a. Specify the desired ref_db value for each background track independently
b. Specify a "master" background track, and then provide an SNR for all other background tracks with respect to the master background.

Thoughts?

**Comment by justinsalamon** [comment]
p.s. - regardless of the direction taken, this would imply a pretty big change to the API, so I expect this to require some pretty detailed planning prior to any implementation.

**Comment by justinsalamon** [comment]
For example, if multiple background tracks are provided, then the SNR of each foreground event would have to be computed relative to the *mixture of backgrounds*. This means the background mix would have to be created first, and then all the foreground events can be added. So it will require not only a large design change to the API but also some re-wiring of the internal generation logic.

**Comment by bensteers** [comment]
For (2), yes I agree that to define SNR properly, you would have to define it relative to a mixture and I think that would be something worth doing at some point. 

So right now, there is technically "support" for multiple background tracks, you can add multiple backgrounds, but they will all have their SNR defined at 0.

I think my misunderstanding was on how you were defining SNR. I saw that Scaper handles multiple background tracks, so SNR at the moment really just equates to some gain offset from ref_db and not "true SNR" for soundscapes with more than 1 bg event. 

Until we can implement SNR for multiple bgs properly, it might make sense to be able to shift this gain seeing as multiple backgrounds are already possible, but I understand the reluctance to add it to the API yet.

**Comment by justinsalamon** [comment]
> I think my misunderstanding was on how you were defining SNR. I saw that Scaper handles multiple background tracks, so SNR at the moment really just equates to some gain offset from ref_db and not "true SNR" for soundscapes with more than 1 bg event.

This is correct. Basically it wasn't written with support for multiple backgrounds in mind, and in fact if you used multiple backgrounds right now the result would be wrong, because every background track would be gain adjusted to `ref_db`, but then they would be summed together, meaning the final background mix would have a greater gain, and then the SNR values of all the foreground sounds would bee wrong.

I think the key to the solution is to support providing an SNR for each background track as you suggested, making it clear that this SNR will be with respect to the `ref_db`. This way you can balance different background tracks by giving them different SNR values with respect to `ref_db`. This also gets rid of the issue of having to define a "master background" track which I think is too clunky.

So, a path towards a fix would look like this I think:
* Allow the user to set an SNR value for the background track(s), which is computed relatively to `ref_db`.
  * Until we determine what to do about default values, this SNR value will have to be specified explicitly.
* In the generation code, first combine all the background tracks. Then compute the LUFS value of the mix `bg_mix_db`. Then adjust the gain of all foreground events with respect to `bg_mix_db` instead of `ref_db`.

This would require major changes to the API, the generation logic, new unit tests and, importantly, changes to the documentation and the examples. It might also have unexpected interactions with the PR for saving individual sound events for source separation.

So I would suggest that we keep thinking about this and discussing it, but I would put a pin in it until the other PRs that are more advanced have been merged.

**Comment by justinsalamon** [comment]
p.s - and as a user I would probably refrain from using multiple background tracks right now, since the resulting soundscape would have wrong SNR values for all the foreground tracks. It's definitely an issue that needs addressing, but for now best to treat it as not supported.

**Comment by bensteers** [comment]
I think that's a good solution, and yeah "master track" is defs too clunky. 

I think there are a few uncontroversial default values that we could add ( `pitch_shift=None, time_stretch=None, source_file=('choose, []), bg.snr=('const', 0)`) and like you said, SNR for background events doesn't make too much sense, so I realllyy don't think it should be made a required field.

**Comment by justinsalamon** [comment]
Another question is whether, once the background tracks are mixed together, to keep them as is (and adjust the gain of foreground events relative to `bg_mix_db`) or, alternatively, to mix the background tracks and then re-adjust the gain of the mix back to `ref_db`, in which case the logic for adjusting the gain for foreground events can remain unchanged. Not sure which option makes more sense.

There definitely are a few uncontroversial default values that could be used, but I wouldn't make this change piecemeal - we should determine default values for all parameters and then add them all in one go.

**Comment by bensteers** [comment]
Can't we use `ref_db` for both? Like mix bgs with `ref_db` and then re-adjust the mix back to `ref_db`. It would leave the current API mostly intact with only the addition of an snr as an optional argument to add_background.

**Comment by justinsalamon** [comment]
Yes, that's what I was suggesting basically :)

**Comment by beasteers** [comment]
Lately, I've been thinking about overlaying multiple background signals, so I figured I'd take a minute to sort out combining backgrounds with relative dBs. 

I've only just begun sketching it out (haven't tested anything) seeing as we need to discuss things.

```python
... 
# Array for storing all tmp files (one for every event)
tmpfiles = []
with _close_temp_files(tmpfiles):
    tmpfiles_bg = []
    with _close_temp_files(tmpfiles_bg):

        # build background events
        for e in ann.data:
            if e.value['role'] == 'background':
                tmpfiles_bg.append(
                    self._generate_event_audio(e, tile=True))


        # shift background db level
        if len(tmpfiles_bg):
            if len(tmpfiles_bg) == 1:
                # nothing to combine, just take the single background
                bg_tmpfile = tmpfiles_bg.pop(-1)

            else:
                # combine multiple background signals
                cmb = sox.Combiner()

                bg_tmpfile = tempfile.NamedTemporaryFile(
                    suffix='.wav', delete=False)
                cmb.build([t.name for t in tmpfiles_bg],
                          bg_tmpfile.name, 'mix')

            # compute LUFS for the overall background audio
            bg_lufs = get_integrated_lufs(bg_tmpfile.name)
            gain = self.ref_db - bg_lufs

            # only apply gain the background if we need to
            if gain:
                # Normalize to specified SNR with respect to reference db
                tfm = sox.Transformer()
                tfm.gain(gain_db=gain, normalize=False)

                tmpfiles.append(
                    tempfile.NamedTemporaryFile(suffix='.wav', delete=False))
                tfm.build(bg_tmpfile.name, tmpfiles[-1].name)
            else:
                tmpfiles.append(bg_tmpfile)

    # create foreground events
    for e in ann.data:
        if e.value['role'] == 'foreground':
            tmpfiles.append(
                self._generate_event_audio(
                    e, fade_in_out=(self.fade_in_len, self.fade_out_len)))
# then combine everything like normal
... 

```

In the process of making this change, I also factored out the code that generates the foreground and background events into a single function `_generate_event_audio` which treats events as role agnostic, but uses parameters to modify behavior. 

So basically, if you allow `time_stretch` and `pitch_shift` to be added to `add_bg(...)` it would apply them to background events, and if not, it still preserves current functionality. 

I'm fine with holding off on merging anything with master, but I might roll with it locally for personal use. 

I can see how this would affect the source separation pull request, seeing as it merges the background into 1 file. In order to get the separate backgrounds as tracks we would need to work out how to get the separated tracks with proper gain applied. In the meantime, seeing as scaper only officially supports a single background track now, we could just output a single mixed background until we get that figured out.

**Comment by beasteers** [comment]
And if you want to know what `_generate_event_audio` would look like:
```python 
def _generate_event_audio(self, e, fade_in_out=False, tile=False):
    # Possibly tile the audio if the source is shorter than the duration of
    # the event (and if the tile flag is set). `tile` is set to False for
    # foreground events.
    source_duration = (sox.file_info.duration(e.value['source_file']))

    ntiles = int(e.value['event_duration'] // source_duration + 1)
    if not tile and ntiles > 1:
        raise ScaperError(
            'Source duration ({:.2f}s) is too short for '
            'event duration ({:.2f}s).'.format(
                source_duration, e.value['event_duration']))

    if ntiles > 1: # Create combiner
        tfm = sox.Combiner()
    else: # Create transformer
        tfm = sox.Transformer()

    # Ensure consistent sampling rate and channels
    tfm.convert(samplerate=self.sr,
                n_channels=self.n_channels,
                bitdepth=None)
    # Trim
    tfm.trim(e.value['source_time'],
             e.value['source_time'] + e.value['event_duration'])

    # Pitch shift
    if e.value['pitch_shift'] is not None:
        tfm.pitch(e.value['pitch_shift'])

    # Time stretch
    if e.value['time_stretch'] is not None:
        tfm.tempo(1.0 / float(e.value['time_stretch']))

    # Apply very short fade in and out
    # (avoid unnatural sound onsets/offsets)
    if fade_in_out:
        fade_in_len, fade_out_len = fade_in_out
        tfm.fade(fade_in_len=fade_in_len,
                 fade_out_len=fade_out_len)


    # PROCESS BEFORE COMPUTING LUFS
    tmpfiles_internal = []
    with _close_temp_files(tmpfiles_internal):
        # create internal tmpfile
        tmpfiles_internal.append(
            tempfile.NamedTemporaryFile(suffix='.wav', delete=False))

        # synthesize concatenated/trimmed background
        if ntiles > 1: 
            tfm.build([e.value['source_file']] * ntiles,
                      tmpfiles_internal[-1].name, 'concatenate')

        # synthesize edited foreground sound event
        else: 
            tfm.build(e.value['source_file'],
                      tmpfiles_internal[-1].name)


        # if time stretched get actual new duration
        if e.value['time_stretch'] is not None:
            event_duration = sox.file_info.duration(tmpfiles_internal[-1].name)
        else:
            event_duration = e.value['event_duration']


        # NOW compute LUFS
        e_lufs = get_integrated_lufs(tmpfiles_internal[-1].name)

        # Normalize to specified SNR with respect to reference db
        tfm = sox.Transformer()
        gain = self.ref_db + e.value['snr'] - e_lufs
        tfm.gain(gain_db=gain, normalize=False)


        # Pad with silence before/after event to match the
        # soundscape duration. Background events implicitly have (0,0) padding
        prepad = e.value['event_time']
        postpad = max(0, self.duration - (e.value['event_time'] + event_duration))
        if prepad or postpad:
            tfm.pad(prepad, postpad)
        

        # Finally save result to a tmp file
        tmpfile = tempfile.NamedTemporaryFile(suffix='.wav', delete=False)
        tfm.build(tmpfiles_internal[-1].name, tmpfile.name)
        
    return tmpfile
```

**Comment by justinsalamon** [comment]
@beasteers it's generally hard for me to evaluate / discuss ideas when presented in the form of code, because it requires a lot of time to go over the logic line by line.

The more I speak with people about it, the more I think supporting multiple background is a subtle issue that needs to be handled with great care. Some things to consider for example:
* Do we concatenate short background files to match the soundscape duration (as is done now), or not?
* Do we support relative differences in dB between background files? What happens when we re-normalize the entire background track to ref_db? 
* Should we worry about clipping when overlaying a large number of background files?

My preferred way to approach this would be to spec out a complete solution, including all corner and edge cases (or at least as many as we can think of) in prose. Once this solution is sketched out, it can be translated into unit tests and code.

I don't have any cycles to work on this right now, but keeping this open for future reference.
